allowed.py

```diff
@@ -524,32 +549,43 @@ def check_file(filename: str, constructs: tuple, check_method_calls: bool) -> No
 
 
 def read_jupyter_notebook(file_contents: str) -> tuple:
-    """Returns a tuple (x, y), where x is a string representation of the
-    code cells, and y is is a mapping of line numbers to cell and line numbers
+    """Returns a tuple (x: str, y: dict), where x is the concatenated source
+    from the code cells and y is is a mapping of line numbers in x to the
+    corrisponding cell and line numbers from the notebook. Negative Keys in
+    y map to a cells with syntax errors.
 
-    Cells with syntax errors and magics will be skipped. Cell numbers
-    corrispond to code cells only.
+    Ipython magics are transformed into valid python if ipython is installed, otherwise
+    cells with magics are mapped for syntax error.
     """
     cell_num, source_line_num = 1, 1
+    syntax_error_num = -1
     line_cell_map = {}
+    source_list, cell_lines = [], []
     jobject = json.loads(file_contents)
-    source, cell_lines = [], []
     for cell in jobject["cells"]:
         if cell["cell_type"] == "code":
             for cell_line in cell["source"]:
                 cell_lines.append(cell_line)
             cell_lines[-1] += "\n"
             try:
-                ast.parse("".join(cell_lines))
+                if IPYTHON_INSTALLED:
+                    ast.parse(Transformer().transform_cell("".join(cell_lines)))
+                else:
+                    ast.parse("".join(cell_lines))
                 for cell_line_num, cell_line in enumerate(cell_lines, start=1):
-                    source.append(cell_line)
+                    source_list.append(cell_line)
                     line_cell_map[source_line_num] = (cell_num, cell_line_num)
                     source_line_num += 1
-            except:
-                pass
+            except SyntaxError:
+                line_cell_map[syntax_error_num] = (cell_num, 1)
+                syntax_error_num -= 1
```

Comment:
You're using the same dictionary for two different purposes with two different keys: line numbers and error numbers. That will lead to hard-to-debug errors. Separate the purposes, e.g. use a simple list to remember which cells have errors, and add arguments to functions if necessary, but don't shoehorn multiple purposes into the same data structure.

Reply:
Yes I can see your point and I will change this. This is probably one of those habits I have  from solving to many leetcode problems and other puzzles where thoughts about longevity of the code and maintainability are often non existent.

Reply:
This has been changed now so that `read_jupyter_notebooks` returns a tuple of 3 objects. This got me thinking about when (or if) there is a time to start thinking about bundling the information about Jupyter notebooks into some kind of class. `read_jupyter_notebooks` could be replaced with the constructor and `get_line` could be encapsulated to. Moreover, there could be a time in the future when we want to allow or disallow individual magics and report the errors back as magics (as apposed to get_ipython() calls) and it might make more sense to have these as methods of some kind of object.

At the same time I appreciate that this is just a small script to check if certain constructs are allowed so this might be complete overkill. what do you think? is this something that could be done in the future?